src/irr/asset/CImageLoaderOpenEXR.cpp

```diff
@@ -41,17 +41,69 @@ namespace irr
 
 			SContext ctx;
 
-			readVersionField(_file, ctx);
-			readHeader(fileName, ctx);
+			// readVersionField(_file, ctx);
+			// readHeader(fileName, ctx);
 
 			Array2D<Rgba> pixels;
 			int width;
 			int height;
 
 			readRgba(fileName, pixels, width, height);
 
-			// TODO put data into buffer, consider OpenEXR 2.0 supported saving layouts 
-			return SAssetBundle{};
+			constexpr uint32_t MAX_PITCH_ALIGNMENT = 8u;									   // OpenGL cannot transfer rows with arbitrary padding
+			auto calcPitchInBlocks = [](uint32_t width, uint32_t blockByteSize) -> uint32_t	   // try with largest alignment first
+			{
+				auto rowByteSize = width * blockByteSize;
+				for (uint32_t _alignment = MAX_PITCH_ALIGNMENT; _alignment > 1u; _alignment >>= 1u)
+				{
+					auto paddedSize = core::alignUp(rowByteSize, _alignment);
+					if (paddedSize % blockByteSize)
+						continue;
+					return paddedSize / blockByteSize;
+				}
+				return width;
+			};
+
+			ICPUImage::SCreationParams params;
+			params.format = EF_R16G16B16A16_SFLOAT;
+			params.type = ICPUImage::ET_2D;;
+			params.flags = static_cast<ICPUImage::E_CREATE_FLAGS>(0u);
+			params.samples = ICPUImage::ESCF_1_BIT;
+			params.extent.width = width;
+			params.extent.height = height;
+			params.extent.depth = 1u;
+			params.mipLevels = 1u;
+			params.arrayLayers = 1u;
+
+			auto& image = ICPUImage::create(std::move(params));
+
+			auto texelBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(image->getImageDataSizeInBytes());
+			auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1u);
+			ICPUImage::SBufferCopy& region = regions->front();
+			//region.imageSubresource.aspectMask = ...; // waits for Vulkan
+			region.imageSubresource.mipLevel = 0u;
+			region.imageSubresource.baseArrayLayer = 0u;
+			region.imageSubresource.layerCount = 1u;
+			region.bufferOffset = 0u;
+			region.bufferRowLength = calcPitchInBlocks(width, getTexelOrBlockBytesize(image->getCreationParameters().format));
+			region.bufferImageHeight = 0u;
+			region.imageOffset = { 0u, 0u, 0u };
+			region.imageExtent = image->getCreationParameters().extent;
+
+			float* dataToSend = _IRR_NEW_ARRAY(float, image->getImageDataSizeInBytes());
```

Comment:
you could do `dataToSend = texelBuffer->getPointer()` instead of wasting a copy.

Reply:
but it won't be a reference pointer, but simply pointer. Won't it crash?

Reply:
as long as the ICPUBuffer lives (the smart pointer doesn't go out of scope), the pointer returned by `getPointer` is valid